src/openms/include/OpenMS/KERNEL/MSExperiment.h

```diff
@@ -336,62 +337,180 @@ namespace OpenMS
         }
-        
-        const MSSpectrum& spectrum = it.getSpectrum();
-        bool has_IM = spectrum.containsIMData();
-        float peak_IM{-1.0f};
-        if (has_IM)
+
+        if (unit != DriftTimeUnit::NONE)
         {
-          const auto& im_data = spectrum.getIMData();
           const Size peak_index = it.getPeakIndex().peak;
-          if (spectrum.getFloatDataArrays()[im_data.first].size() == spectrum.size())
-          {
-            peak_IM = spectrum.getFloatDataArrays()[im_data.first][peak_index];
-          }          
+          ion_mobility.back().push_back(im[peak_index]);
+        }
+        else
+        {
+          ion_mobility.back().push_back(-1.0);
         }
-        ion_mobility.back().push_back(peak_IM);
         mz.back().push_back((float)it->getMZ());
         intensity.back().push_back(it->getIntensity());
       }
     }
 
+    /* @brief Retrieves the peak data in the given mz-rt range and store in separate arrays.
+     * 
+     * For fast pyOpenMS access to MS1 peak data in format: [rt, mz, intensity]
+     * 
+     * @param min_rt The minimum retention time.
+     * @param max_rt The maximum retention time.
+     * @param min_mz The minimum m/z value.
+     * @param max_mz The maximum m/z value.
+     * @param ms_level The MS level of the spectra to consider.
+     * @param rt The vector to store the retention times in.
+     * @param mz The vector to store the m/z values in.
+     * @param intensity The vector to store the intensities in.
+    */    
+    void get2DPeakData(
+      CoordinateType min_rt,
+      CoordinateType max_rt,
+      CoordinateType min_mz,
+      CoordinateType max_mz,
+      std::vector<float>& rt,
+      std::vector<float>& mz,
+      std::vector<float>& intensity) const
+    {
+      for (auto it = areaBeginConst(min_rt, max_rt, min_mz, max_mz); it != areaEndConst(); ++it)
+      {
+        rt.push_back((float)it.getRT());
+        mz.push_back((float)it->getMZ());
+        intensity.push_back(it->getIntensity());
+      }
+    }
+
+
+    /* @brief Retrieves the peak data in the given mz-rt range and store in separate arrays.
+     * 
+     * For fast pyOpenMS access to MS1 peak data in format: [rt, mz, intensity, ion mobility]
+     * 
+     * @param min_rt The minimum retention time.
+     * @param max_rt The maximum retention time.
+     * @param min_mz The minimum m/z value.
+     * @param max_mz The maximum m/z value.
+     * @param ms_level The MS level of the spectra to consider.
+     * @param rt The vector to store the retention times in.
+     * @param mz The vector to store the m/z values in.
+     * @param intensity The vector to store the intensities in.
+    */
+    void get2DPeakDataIM(
+      CoordinateType min_rt,
+      CoordinateType max_rt,
+      CoordinateType min_mz,
+      CoordinateType max_mz,
+      Size ms_level,
+      std::vector<float>& rt,
+      std::vector<float>& mz,
+      std::vector<float>& intensity,
```

Comment:
Possibly should have the option here to also filter by ion mobility

Reply:
Agree. Not sure if it can be done in an efficient way currently (linear search?).
Should we add an additional function with and without IM filtering?
Not sure if you can pass Infinity from python to C and enable/disable IM filtering based on that.

Reply:
I think this should be possible as both use IEEE 754 floating-point rules

Reply:
I will keep this feature request for the next iteration.